### ISSUE TITLE
Initial navigation blocked if website has service worker that enables navigation preload

### ISSUE THREAD

**Issue by steffenweber** [uBlockOrigin/uMatrix-issues#156]
### Prerequisites

- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
    - [x] I also searched the existing issues at <https://github.com/gorhill/uMatrix/issues>
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uMatrix](https://old.reddit.com/r/uMatrix/)
- I tried to reproduce the issue when...
    - [x] uMatrix is the only extension
    - [x] uMatrix with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uMatrix
- [x] I checked the [documentation](https://github.com/gorhill/uMatrix/wiki) to understand that the issue I report is not a normal behavior
- [x] I used the logger to rule out that the issue is caused by my ruleset

### Description

If a website has a service worker that enables [Navigation Preload](https://developers.google.com/web/updates/2017/02/navigation-preload) then such navigation preload requests are blocked by uMatrix. This then causes the page to fail to load.

### A specific URL where the issue occurs

https://www.computerbase.de/

### Steps to Reproduce

1. Open https://www.computerbase.de/ to initialize the service worker
2. In the same tab, do an omnibox search for "computerbase forum"
3. On the Google search results, click the first link (should point to https://www.computerbase.de/forum/)

What should happen: The navigation preload request for https://www.computerbase.de/forum/ should pass the checks of uMatrix. The page should load fine.

What actually happens: The navigation preload request for https://www.computerbase.de/forum/ is blocked by uMatrix (see the screenshot below). This causes our service worker to show an "offline" fallback/error page.

(Reproducing this issue is not 100% reliable. Afaik, Chrome only sends navigation preload requests when the service worker is not already running. Maybe you need more than one try to reproduce this issue. You can try to manually stop the computerbase.de service worker in the "Application" tab of the Chrome developer tools between steps 2 and 3.)

This issue was originally reported by one of our users here: https://www.computerbase.de/forum/threads/login-probleme-seit-14-mai-der-aktive-benutzer-hat-sich-geaendert.1872153/post-22667228

### Ruleset

(I've used a fresh Chrome profile and didn't add/delete any rules. )

### Supporting evidence

![Screenshot from 2019-05-21 16-03-24-fs8](https://user-images.githubusercontent.com/1138129/58102876-54f9e780-7be2-11e9-9482-75553887c7fc.png)

![Screenshot from 2019-05-21 16-06-40-fs8](https://user-images.githubusercontent.com/1138129/58102985-7c50b480-7be2-11e9-9406-cd53a0c2dab1.png)

### Your environment

* uMatrix version: 1.3.16 / 1.3.17rc1
* Browser Name and version: Chrome 74.0.3729.131 / 75.0.3770.27
* Operating System and version: Gentoo Linux (x86_64)

**Comment by uBlock-user** [comment]
> What actually happens: The navigation preload request for https://www.computerbase.de/forum/ is blocked by uMatrix (see the screenshot below). This causes our service worker to show an "offline" fallback/error page.

Site loads fine, no such offline error page is shown. Secondly, no such "service worker navigation preload request" error appears in the console either.

**Comment by gorhill** [comment]
I am able to reproduce.

**Comment by uBlock-user** [comment]
As per `chrome://serviceworker-internals`, _navigation preload enabled_ is set to `false`, this is why it seems I can't reproduce.

![](https://i.gyazo.com/58a2889e0b0800aa895f732805e4e09d.png)

**Comment by steffenweber** [comment]
Screen capture of the issue in Chrome 68 Beta and uMatrix 1.3.17rc1: https://youtu.be/r7qKTpD-fc0

(I've just read that you are now able to reproduce the issue but maybe the video can still be useful. :slightly_smiling_face:)

**Comment by gorhill** [comment]
Unfortunately, Chrome is passing bad information to uMatrix:

![a](https://user-images.githubusercontent.com/585534/58105255-2e24bc80-7bb4-11e9-86c6-7389449dab3d.png)

It's telling uMatrix that the network request of type `other` to `https://www.computerbase.de/forum/` originates from `https://www.google.com`. There is nothing in there that could be used to fall back from this bad information. The only solution I can think of for a uMatrix user is to create a rule from the logger such as:

![a](https://user-images.githubusercontent.com/585534/58105745-fc602580-7bb4-11e9-9a55-57410a09d979.png)

**Comment by mikhaelkh** [comment]
If I search `uMatrix` in DuckDuckGo then click on `https://github.com/gorhill/uMatrix`, COOKIES are blocked and I'm not logged in.
I have 
```
* * cookie block
* 1st-party cookie allow
```
It seems to not be specific to Service Worker with Navigation Preload.

**Comment by uBlock-user** [comment]
Ah yes, I can reproduce that. Seems to be occuring on very first load only and from the past two days.

**Comment by uBlock-user** [comment]
@mikhaelkh I believe [https://github.com/gorhill/uMatrix/commit/901c325eabcc2736119ffef46ffc2d978107c969](documentUrl commit) appears to be the cause.

**Comment by gorhill** [comment]
That commit just made the browser bug more prominent, it's happening also because Chrome is feeding bad information to uBO. However I think I can mitigate this one because the request type is `main_frame`, and in that case it just make sense to have the originator set the same as the request URL. This won't fix the issue here, but this will fix that spurious cookie deletion one.

**Comment by uBlock-user** [comment]
> This won't fix the issue here, but this will fix that spurious cookie deletion one.

Thanks, this spurious deletion is literally trolling me on youtube and github where I'm losing my cookie based settings randomly every now and then and me having to set them again and again.

**Comment by gorhill** [comment]
Closing since there is nothing which can be done to work around this Chrome issue.

**Comment by steffenweber** [comment]
You said that "Chrome is passing bad information to uMatrix". This sounds like a bug in Chrome to me. Could you report this issue at https://crbug.com?

I'm not familiar with the code of uMatrix or the WebExtensions API.

**Comment by steffenweber** [comment]
I have filed a Chrome bug report: https://bugs.chromium.org/p/chromium/issues/detail?id=965960

Please leave a comment if I've missed something.
